### DISCUSSION TITLE
tutorials/lift-off-part1/10-the-usequery-hook

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#36] [course feedback]
# tutorials/lift-off-part1/10-the-usequery-hook

https://www.apollographql.com/tutorials/lift-off-part1/10-the-usequery-hook

**Comment by jkolezyn** [comment]
Awesome tutorial. I wish all tutorials were made like this. Thanks a lot!

**Comment by bankelal** [comment]
Thank you so much for this great Tutorials. 

Request you to please fix the video play back issue. If I pause and play again, it does not load at all.

I am getting this error after copying the complete solution in tracks.js.

The previous chapter worked great and I was able to see the list. 


Module parse failed: Unexpected token (40:10)
You may need an appropriate loader to handle this file type.
|       columnNumber: 9
|     }
>   }, data?.tracksForHome?.map(function (track, index) {
|     return /*#__PURE__*/React.createElement(TrackCard, {
|       key: track.id,

**Reply by bankelal** [reply]
@lizhennessy please help!

**Reply by lizhennessy** [reply]
Hi there @bankelal - I'm so sorry for the delay in my response to you! This is an interesting error. Can you copy the code for your entire file? It looks like there might be a syntax error somewhere. Looking forward to your response!

**Comment by angelotheman** [comment]
Wow, this was very awesome. Building with graphql is so nice.

**Comment by HectorDevelops** [comment]
Thank you for the overview!

I do have a question - after switching to client and adding the query code (up to 1:49) the pages outputs the "Error! Load failed" output. 
Any workaround that?

**Reply by mabuyo** [reply]
Hey there @HectorDevelops! There could be an issue with your server unable to return the data. It's difficult to troubleshoot without more information. Would you be able to share your repo and I can try to reproduce?

**Reply by HectorDevelops** [reply]
It ended up being an issue with my server and was able to fix it!
Thanks for the quick turn around time.

**Reply by mabuyo** [reply]
Glad to hear you figured it out! :)

**Comment by VRedondo** [comment]
Thanks for the tutorial, it's awesome

I would like to know why even thought I generate the types based on the graphql schema when doing 

```js 
useQuery<Query>(TRACKS);
```

Should I always mark the generic type as Query? or is there a way to default all useQuery to Query type generated ? I know this is more a TS thing but I was curious 

Thanks

**Reply by BlenderDude** [reply]
While the `useQuery` type does specify that it needs a type parameter of `<Query>`, if the right argument is passed to `useQuery`, it is able to automatically calculate it for you! Part of the magic of graphql codegen is in the typing of the `gql()` function. After running the codegen command, when it sees a query, instead of returning a `string`, it instead returns this `TypedDocumentNode<Result, Variables>` type. This type contains the `Result` and `Variables` information such that your `useQuery` function has all of the proper typing. If your codegen is set up properly, hovering over `TRACKS` will reveal `TypedDocumentNode` with `Result` and `Variables` filled in to match the `TRACKS` query.

Given that, useQuery(TRACKS)  is all that is necessary!

**Comment by Sharonah8** [comment]
The cards are not displaying on my end.  Just the plain initial homepage that has the navbar and the footer.

**Reply by mabuyo** [reply]
Sorry to hear that @Sharonah8! If you push up your commits to the same repo https://github.com/Sharonah8/apollo-graph-ql-basics, I can take a look. Difficult to troubleshoot without it!

**Reply by Sharonah8** [reply]
Hey. Sorry for late response. I've pushed

**Comment by gardenboi** [comment]
the spinner doesn't show up with ctrl+f5 

in the 
```
const QueryResult: React.FC<PropsWithChildren<QueryResultProps>> = ({
  loading,
  error,
  data,
  children,
}): React.ReactElement<any, any> | null => {
  if (error) {
    return <p>ERROR: {error.message}</p>;
  }
  if (loading) {
    return (
      <SpinnerContainer>
        <LoadingSpinner data-testid="spinner" size="large" theme="grayscale" />
      </SpinnerContainer>
    );
  }
  if (!data) {
    return <p>Nothing to show...</p>;
  }
  if (data) {
    return <>{children}</>;
  }
};
```
I had to add 

```
return <p>nothing</p>
```

at the end of the component because typescript will ask for a default return statement otherwise it gives error

**Reply by mabuyo** [reply]
Hey @gardenboi ! I wasn't able to reproduce the issue with the loading spinner. The loading state is quite fast, so it's possible you may have missed it. But even simulating slow network speeds, I was able to see the spinner.

However, I checked the starter code and saw that the code snippet is slightly different from what we showed in the lesson contents. The lesson will be updated shortly to reflect [what we have in the starter code](https://github.com/apollographql/odyssey-lift-off-part1/blob/typescript/final/client/src/components/query-result.tsx), which does contain a default return statement. (Although you shouldn't need to touch the `QueryResult` component!)

Hope that helps!

**Comment by BruceRoyce** [comment]
### Why did you get 8 cards?!
When you first implemented the `TrackCard` component in the map loop, you got 8 cards printed _(see **@3:26** of the video, or the **second screenshot** in the description below the video)_. That was changed down to the expected 6 cards after you implemented the `QueryResult` component that should have only improved the visuals.
Why is that? 
We know the `schema` explicitly set up to return a list of 6 results. So where the extra 2 cards are from?

**Reply by BruceRoyce** [reply]
the mock is setup for an array of 6 (not the schema)

**Reply by mabuyo** [reply]
Hey there @BruceRoyce, good catch! I believe the video is showing an older version of the course. At the time of recording, the mocks for the `tracksForHome` array returned a list of tracks with length between 6 and 9 (a random number between that was selected whenever the server returned data back). That's why you see the number of cards change in the UI. 

However, in newer versions of the course, we opted to go with the a list of just 6 results but the screencast was left alone. Sorry for the confusion!

**Reply by BruceRoyce** [reply]
thank you for clarification @mabuyo

**Comment by neogeek13** [comment]
Just about wrapping this excellent course up and noticed a tiny typo on this last lesson:

"and more accurately describe of data the track prop should provide."

Looking forward to the next course!

**Reply by mabuyo** [reply]
Ahh good catch, will fix shortly! Glad you enjoyed!

**Comment by LachPawel** [comment]
Great job on this tutorial, thank you! :)

**Comment by dhawal-bajaj** [comment]
In the final code of the <code>tracks.js</code> file, the below <code>import</code> statement is missing:
<code>import QueryResult from '../components/query-result';</code>

**Comment by laxmipatri** [comment]
Getting ERROR: Response not successful: Received status code 400. Please help.

**Reply by lizhennessy** [reply]
Hey there @laxmipatri ! Can you confirm that your server is still running on port 4000 and that you're able to query it in Sandbox by going to http://localhost:4000? If all of that looks good, are you seeing any particular error messages in the console?

**Comment by kaliaparijat** [comment]
There seems to be a disparity in the article between writing a query with GQL tags and suggesting using the codegen generated Graphql function as has been advocated here: 

https://www.apollographql.com/tutorials/lift-off-part1/09-codegen#:~:text=%40graphql%2Dcodegen%2Fcli%20is,to%20use%20on%20the%20frontend. 

The tutorial continues to use the gql tags from graphql. Am I missing something? I ask because my application doesn't like the generated function and complains about a type mismatch

**Reply by lizhennessy** [reply]
Hi there @kaliaparijat ! We use the GraphQL codegen `gql` tag in the TypeScript version of the course, and the `gql` tag from `graphql-tag` in the JavaScript version of the course. Which version were you hoping to complete? (You can change versions in the dropdown at the top of the page!) Currently our videos only cover the JavaScript version.

**Reply by kaliaparijat** [reply]
Thanks @lizhennessy : I was able to view the tutorial in typescript and you are right, it's using the codegen function as indicated in the tutorial. My bad!

**Reply by lizhennessy** [reply]
Not to worry @kaliaparijat ! Glad we've got it solved! 🔥

**Reply by kaliaparijat** [reply]
Thanks @lizhennessy! Unfortunately, only partially so. When I run codegen, that generates the gql file as expected but it ends up generating something like this in that gql file:

``` *
 * The query argument is unknown!
 * Please regenerate the types.
 */
export function graphql(source: string): unknown;

{...along with 14 other graphql function overloads }
```

That doesn't work well with Apollo useQuery hooks which are expecting queries that are of type documentNode whereas the function definition here has an unknown type. 

Any idea why this could happen?

**Reply by lizhennessy** [reply]
Hmm that's an interesting error, @kaliaparijat ! Do you have your code pushed up somewhere that I can access to take a look?

**Reply by kaliaparijat** [reply]
Thanks for responding, @lizhennessy : I can't upload the repository to anywhere public, but I can share relevant code:

the config file:
```
const config: CodegenConfig = {
  overwrite: true,
  schema: path.join(__dirname, "./schema.graphql"),
  documents: ["src/**/*.tsx", "src/**/*.ts"], // All queries are under "src/queries/*.ts". 
  generates: {
    "src/queries/__generated__/": {
      preset: "client",
      plugins: ["typescript"],  // same result with or without plugin
    },
    "./graphql.schema.json": {
      plugins: ["introspection"],
    },
  },
};
```

An example query file:
```
 /* 
 this imports the first Graphql overload function that has an unknown type, for each of the queries there is a generated
* graphql function that doesn't have a type unknown, but of a typeof documents (map of all graphql operations in the project) 
*/
import { graphql } from "./__generated__"; 

export const GET_DATA = graphql(`
  query getData($Criteria: CriteriaInputType) {
    dataSet(criteria: $Criteria) {
          pagedResult {
              ...FragmentTypeA
              ...FragmentTypeB
          }
    },
    validations {
      key
      value
    }
  }
  ...FragmentTypeA,
  ...FragmentTypeB
`);```

**Comment by davidemmolino** [comment]
Hello, I'm getting this rendered to the page `Error! Response not successful: Received status code 404`

Confirming that my server is still running on port 4000 and that I'm able to run the query in the sandbox

The browser console shows error 404 not found.

This is what the tracks.js file looks like 
```
import React from 'react';
import { Layout } from '../components';
import { useQuery, gql } from "@apollo/client";
import TrackCard from "../containers/track-card";

const TRACKS = gql`
query GetTracks {
  tracksForHome {
    id
    title
    author {
      id
      name
      photo
    }
    thumbnail
    length
    modulesCount
  }
}
`
/**
 * Tracks Page is the Catstronauts home page.
 * We display a grid of tracks fetched with useQuery with the TRACKS query
 */
const Tracks = () => {
  const { loading, error, data } = useQuery(TRACKS);
  if (loading) return "Loading...";
  if (error) return `Error! ${error.message}`;
  return <Layout grid>{JSON.stringify(data)}</Layout>
};

export default Tracks;

```

**Reply by lizhennessy** [reply]
Hi there @davidemmolino! Thanks for reaching out - do you happen to have a repo where I can check out your code? With this code in the `tracks.js` file you should at the very least be seeing the `JSON` stringified data, so I suspect the trouble might be elsewhere!

**Comment by onesamket** [comment]
it was amazing thanks  a lot 🙏

**Comment by horrorkist** [comment]
thanks for the legend tutorial

I have a question though, I expected that i could do something like below

```
const Tracks = () => {
  const { loading, error, data } = useQuery<GetTracksResult>(TRACKS); 
  if (loading) return "Loading...";
  if (error) return `Error! ${error.message}`;
  return <Layout grid>{JSON.stringify(data)}</Layout>
};
```
but the data has gettracks field then there's getTracksResult in it.
So I am doing
```
const Tracks = () => {
  const { loading, error, data } = useQuery<{getTracks: GetTracksResult}>(TRACKS);
  if (loading) return "Loading...";
  if (error) return `Error! ${error.message}`;
  return <Layout grid>{JSON.stringify(data)}</Layout>
};
```

Is there any other way to achieve what i'm trying to do?
or am i doing something wrong?

big thanks.

**Reply by horrorkist** [reply]
I figured it out.
I was actually triyng to do login mutation and found LoginMutation type

**Comment by mandyparson** [comment]
Excellent course! Thank you so much for all the help.
Just a quick FYI: Your last code block on this page has some issues. 1) It's missing the QueryResult import and 2) your keys aren't unique.

**Comment by le-el** [comment]
thanks alot

**Comment by lycha0206** [comment]
I am wondering how come you don't name the files for TrackCard and QueryResult as starting with capital letters?  I thought that's the convention: component filenames start with capital letters.

**Reply by lizhennessy** [reply]
Hey there @lycha0206! In this case, it's really a matter of personal preference. ReactJS doesn't enforce one convention or another; it's happy just as long as the component that we actually export from the file is named with a capital letter. :)

**Comment by lycha0206** [comment]
It is a big loop. I remember in 2007 and 2008, I just told my backend team to get me a URL that I can fetch with all cats info and authors info, and they'd give me that in an hour.  Now I have to configure and configure for an hour and finally I have it.  If I can only do that last part: just in the client, list what I want, and run a script to generate the code for the backend, that will be a lot better. Or in the future, I wonder if you can have some kind of AI, and you can type in:  Just add to the backend and frontend -- I need all cats info and the authors info. And in 3 seconds, all files are generated for you.

**Reply by lycha0206** [reply]
I said it is a big loop because in 2010, most companies say: no no no, we won't ask the backend team for that URL. We will do it the RESTful way, and everybody treated RESTful as the Holy Bible. And then later on, it is the GraphQL way, which is basically back to 2007 and 2008.

**Comment by mshah100** [comment]
I have my track.js file as follows, my application gives the following error and does not load.  What a I doing wrong?  Struggling on this for the last week.  As soon as I add useQuery I get the error.

@apollo_client.js?v=8cf7e5cd:182 Uncaught Invariant Violation: An error occurred! For more details, see the full error text at https://go.apollo.dev/c/err#%7B%22version%22%3A%223.8.8%22%2C%22message%22%3A49%2C%22args%22%3A%5B%5D%7D
    at new InvariantError2 (http://localhost:3000/node_modules/.vite/deps/@apollo_client.js?v=8cf7e5cd:182:26)
    at invariant (http://localhost:3000/node_modules/.vite/deps/@apollo_client.js?v=8cf7e5cd:193:11)
    at invariant3 (http://localhost:3000/node_modules/.vite/deps/@apollo_client.js?v=8cf7e5cd:291:5)
    at useApolloClient (http://localhost:3000/node_modules/.vite/deps/@apollo_client.js?v=8cf7e5cd:17185:3)
    at useQuery (http://localhost:3000/node_modules/.vite/deps/@apollo_client.js?v=8cf7e5cd:17342:27)
    at Tracks (http://localhost:3000/src/pages/tracks.js?t=1704408886783:21:36)
    at renderWithHooks (http://localhost:3000/node_modules/.vite/deps/react-dom.js?v=8cf7e5cd:11344:26)
    at mountIndeterminateComponent (http://localhost:3000/node_modules/.vite/deps/react-dom.js?v=8cf7e5cd:13417:21)
    at beginWork (http://localhost:3000/node_modules/.vite/deps/react-dom.js?v=8cf7e5cd:14151:22)
    at HTMLUnknownElement.callCallback2 (http://localhost:3000/node_modules/.vite/deps/react-dom.js?v=8cf7e5cd:1077:22)
InvariantError2 @ @apollo_client.js?v=8cf7e5cd:182
invariant @ @apollo_client.js?v=8cf7e5cd:193
invariant3 @ @apollo_client.js?v=8cf7e5cd:291
useApolloClient @ @apollo_client.js?v=8cf7e5cd:17185
useQuery @ @apollo_client.js?v=8cf7e5cd:17342
Tracks @ tracks.js:30
renderWithHooks @ react-dom.js?v=8cf7e5cd:11344
mountIndeterminateComponent @ react-dom.js?v=8cf7e5cd:13417
beginWork @ react-dom.js?v=8cf7e5cd:14151
callCallback2 @ react-dom.js?v=8cf7e5cd:1077
invokeGuardedCallbackDev @ react-dom.js?v=8cf7e5cd:1102
invokeGuardedCallback @ react-dom.js?v=8cf7e5cd:1133
beginWork$1 @ react-dom.js?v=8cf7e5cd:17141
performUnitOfWork @ react-dom.js?v=8cf7e5cd:16465
workLoopSync @ react-dom.js?v=8cf7e5cd:16450
performSyncWorkOnRoot @ react-dom.js?v=8cf7e5cd:16187
scheduleUpdateOnFiber @ react-dom.js?v=8cf7e5cd:15842
updateContainer @ react-dom.js?v=8cf7e5cd:18000
(anonymous) @ react-dom.js?v=8cf7e5cd:18290
unbatchedUpdates @ react-dom.js?v=8cf7e5cd:16294
legacyRenderSubtreeIntoContainer @ react-dom.js?v=8cf7e5cd:18289
render @ react-dom.js?v=8cf7e5cd:18352
(anonymous) @ index.js:14
Show 21 more frames
Show less
index.js:14 The above error occurred in the <Tracks> component:
    in Tracks (created by Pages)
    in RenderedRoute (created by Routes)
    in Routes (created by Pages)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by Pages)
    in Pages
    in StrictMode


import React from 'react';
import { Layout } from '../components';
import { useQuery, gql } from '@apollo/client';

/*import TrackCard from '../containers/track-card';

/** TRACKS query to retrieve all tracks */
const TRACKS = gql`
  query GetTracks {
    tracksForHome {
      id
      title
      author {
        id
        name
        photo
      }
      thumbnail
      length
      modulesCount
    }
  }
`

/**
 * Tracks Page is the Catstronauts home page.
 * We display a grid of tracks fetched with useQuery with the TRACKS query
 */
const Tracks = () => {
  const { loading, error, data } = useQuery(TRACKS);

  if (loading) return "Loading...";

  if (error) return `Error! ${error.message}`;

  return <Layout grid>{JSON.stringify(data)}</Layout>;
};

export default Tracks;

**Reply by mabuyo** [reply]
Hey there @mshah100 ! One thing I notice is this line:

```js
return {JSON.stringify(data)};
```

At this point in the lesson, it should be:

```js
  return <Layout grid>{JSON.stringify(data)}</Layout>;
```

Can we try that fix? If it still isn't working, it's difficult for me to decipher the code because of the formatting issues, so if you could fix the formatting, I can try to reproduce on my end.

**Reply by mshah100** [reply]
Here is the code changing it to include the Layout tag does not help and is not the cause of the issue:

import React from 'react';
import { Layout } from '../components';
import { useQuery, gql } from '@apollo/client';

/*import TrackCard from '../containers/track-card';

/** TRACKS query to retrieve all tracks */
const TRACKS = gqlquery GetTracks { tracksForHome { id title author { id name photo } thumbnail length modulesCount } }

/**

Tracks Page is the Catstronauts home page.
We display a grid of tracks fetched with useQuery with the TRACKS query
*/
const Tracks = () => {
const { loading, error, data } = useQuery(TRACKS);
if (loading) return "Loading...";

if (error) return `Error! ${error.message}`;

return <Layout grid>{JSON.stringify(data)}</Layout>;

};

export default Tracks;

**Reply by mabuyo** [reply]
The formatting here is still difficult to decipher, if you could use the [triple backticks Markdown syntax](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code), that would be very helpful!

Right now, it looks like there are some lines that have been commented out, starting from:

```
/*import TrackCard from '../containers/track-card';
```
Which might be causing issues.

Another thing you can try is copy-pasting the last code block in the lesson, which is the final state of the code.

**Reply by mshah100** [reply]
I copied the final state of the code from the final / client/ folder still get the same error:

import React from 'react';
import { useQuery, gql } from '@apollo/client';
import TrackCard from '../containers/track-card';
import { Layout, QueryResult } from '../components';

/** TRACKS gql query to retrieve all tracks */
```
const TRACKS = gql`
  query getTracks {
    tracksForHome {
      id
      title
      thumbnail
      length
      modulesCount
      author {
        name
        photo
      }
    }
  }
`;

/**
 * Tracks Page is the Catstronauts home page.
 * We display a grid of tracks fetched with useQuery with the TRACKS query
 */
const Tracks = () => {
  const { loading, error, data } = useQuery(TRACKS);

  return (
    <Layout grid>
      <QueryResult error={error} loading={loading} data={data}>
        {data?.tracksForHome?.map((track, index) => (
          <TrackCard key={track.id} track={track} />
        ))}
      </QueryResult>
    </Layout>
  );
};

export default Tracks;
```

**Reply by mabuyo** [reply]
Thanks for clarifying! When I click on the link in your error, it leads me to [this page](https://www.apollographql.com/docs/react/errors/#%7B%22version%22%3A%223.8.8%22%2C%22message%22%3A49%2C%22args%22%3A%5B%5D%7D), where the error message says:

> Could not find "client" in the context or passed in as an option. Wrap the root component in an <ApolloProvider>, or pass an ApolloClient instance in via options.

In your `client/index.js` file, have you [configured the Apollo Client package](https://www.apollographql.com/tutorials/lift-off-part1/08-apollo-client-setup) correctly? Could you copy-paste the code you have there?

If you have a repo ready with your code, I'm also happy to take a look there!

**Comment by talwar** [comment]
Here is my index.js file under client/src/pages/index.js 

import React from 'react';
import { BrowserRouter, Routes, Route } from 'react-router-dom';
/** importing our pages */
import Tracks from './tracks';

export default function Pages() {
  return (
    <BrowserRouter>
      <Routes>
        <Route element={<Tracks />} path="/" />
      </Routes>
    </BrowserRouter>
  );
}

Here is my track.js under client/src/pages/tracks.js

import React from 'react';
import { Layout } from '../components';
import { useQuery, gql } from "@apollo/client";
import TrackCard from "../containers/track-card";

const TRACKS = gql`
query GetTracks {
  tracksForHome {
    id
    title
    author {
      id
      name
      photo
    }
    thumbnail
    length
    modulesCount
  }
}
`
/**
 * Tracks Page is the Catstronauts home page.
 * We display a grid of tracks fetched with useQuery with the TRACKS query
 */
const Tracks = () => {
  const { loading, error, data } = useQuery(TRACKS);
  if (loading) return "Loading...";
  if (error) return `Error! ${error.message}`;
  return <Layout grid>{JSON.stringify(data)}</Layout>;
};

export default Tracks;

I am getting nothing when I run localhost:3000 It is a blank page.

**Comment by pavel-kalmykov** [comment]
Quick question: are you sure we need to have the `import React from "react";` import in the `tracks.{jsx,tsx}` file? I don't see it being used anywhere (at least in this section).

**Comment by sol-warrior** [comment]
# The useQuery Hook

- The most important part is useQuery()
 
```
const QUERYSPACE = gql(`   
          query GetQName{
                     id
                     title
            }
        `)

const {loading, error,data} = useQuery(QUERYSPACE)


# Finally Done it 💪💪💪

**Comment by zhaoou00** [comment]
I followed all the steps. The server side works fine. The client side just show a blank page on the browser. 
There is NO error message. Just white blank.

I tried to comment out "useQuery(TRACKS)" related code and put 
return ( <Layout grid>dummy    </Layout>  )
instead in tracks.js Then "dummy" shows up on the browser fine.

It seems that as soon as I enable  "useQuery(TRACKS)", the browser turns white blank. 

What pissed me off is that there is no error message telling me where to look.

**Reply by lizhennessy** [reply]
Hey there @zhaoou00 - I'm happy to take a look and help troubleshoot. Do you have your code pushed up to a repository I could check out? I'm curious if there was anything in the network tab that might have indicated what went wrong.

**Reply by zhaoou00** [reply]
It turns out that I was using the default client/src/index.js The default file doesn't have ApolloProvider in it. After I put ApolloProvider in there it starts to work.

<ApolloProvider client={client}>
    <GlobalStyles />
    <Pages />
  </ApolloProvider>

Still it is very annoying that it fails without an error message. There is no hint pointing to this seemingly unrelated place.

**Reply by lizhennessy** [reply]
Hi @zhaoou00! We were able to reproduce on our side and get the blank screen. But it looks like there's a big error in the console that shows what went wrong, and links to the [Apollo Client documentation](https://www.apollographql.com/docs/react/errors#%7B%22version%22%3A%223.10.1%22%2C%22message%22%3A49%2C%22args%22%3A%5B%5D%7D).

<img width="1273" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/95302380/8ebb6e5f-cb99-48db-ad3b-51b45bf50279">

Hopefully this can help in the future!

**Comment by NabsEMan** [comment]
Hi, I have the same problem as zhaoou00.

**Reply by lizhennessy** [reply]
Hi there @NabsEMan ! Do you see the following error in your console?

<img width="1273" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/95302380/8ebb6e5f-cb99-48db-ad3b-51b45bf50279">

If so, you might have a problem with your `ApolloProvider`. It might not be wrapping your app, or you might not have passed the `client` instance to it. If you'd like to paste the contents of your file here, or link me to your repo, I'm happy to troubleshoot further!

**Comment by julillermo** [comment]
Good Day!

I have been following along with the tutorial. I have reached up to the point where the object should be rendered as a string on the client-side on the browser. However, I seem to be blocked by CORS and some source map errors. How may I overcome this?

I received the following errors:
* `Cross-Origin Request Blocked: The Same Origin Policy disallows reading the remote resource at https://localhost:4000/. (Reason: CORS request did not succeed). Status code: (null).`
* `Source map error: request failed with status 404
Resource URL: null
Source Map URL: installHook.js.map`
* `Source map error: No sources are declared in this source map.
Resource URL: http://localhost:3000/node_modules/.vite/deps/chunk-2GTGKKMZ.js?v=51c29799
Source Map URL: chunk-2GTGKKMZ.js.map`
* `Source map error: No sources are declared in this source map.
Resource URL: http://localhost:3000/node_modules/.vite/deps/react.js?v=51c29799
Source Map URL: react.js.map`
* `Source map error: No sources are declared in this source map.
Resource URL: http://localhost:3000/node_modules/.vite/deps/@emotion_core.js?v=51c29799
Source Map URL: @emotion_core.js.map`
* `Source map error: No sources are declared in this source map.
Resource URL: http://localhost:3000/node_modules/.vite/deps/@apollo_space-kit_Loaders_LoadingSpinner.js?v=51c29799
Source Map URL: @apollo_space-kit_Loaders_LoadingSpinner.js.map`
* `Source map error: request failed with status 404
Resource URL: null
Source Map URL: backendManager.js.map`
* `Source map error: request failed with status 404
Resource URL: null
Source Map URL: react_devtools_backend_compact.js.map`

For additional context, I believe that the server works on `http://localhost:4000/` and I can query through the provided GUI

Thank you  in advance!

**Reply by lizhennessy** [reply]
Hello @julillermo ! Happy to take a look and help troubleshoot. Are you following the TypeScript or the JavaScript version of the course?

**Reply by julillermo** [reply]
Hello again!

I redid the tutorial more closely, and I was able to get it to work! I must have missed something along the way. Thank you!

**Comment by bocarw121** [comment]
Amazing tutorial, I learnt a lot. Thanks!

**Comment by Ceheiss** [comment]
Amazing and clear course, it has helped tons to understand some previously unexplored aspects of the codebase I'm working on. Thanks and on to the next one!

**Comment by alfi-dim** [comment]
First, Thanks for the course! It really help me to understand more about GraphQL and Apollo.
But there's a small issue, looks like your url for the author photo and thumbnail is down as i got the broken image icon on the card (sorry, looks like i can't upload an image here), i tested your url too on the browser, and it give me this error

```
No webpage was found for the web address: https://res.cloudinary.com/dety84pbu/image/upload/v1606816219/kitty-veyron-sm_mctf3c.jpg
HTTP ERROR 404
```

**Reply by lizhennessy** [reply]
Ah thank you for letting us know @alfi-dim ! We will take a look to fix this. (Glad that the course really helped too! 😊 )

**Reply by JAndresGarcia** [reply]
+1

**Comment by AndriyYer** [comment]
+1 to the images no longer being hosted at that address. No big deal, I just found some cats online (:

Incredible tutorial, helped me prep for an interview at a company that uses nearly this exact stack.

**Comment by testqual** [comment]
I have tried this and only get white page. Server is working fine. These error are in the console . 
Uncaught Invariant Violation: An error occurred! For more details, see the full error text at https://go.apollo.dev/c/err#%7B%22version%22%3A%223.12.2%22%2C%22message%22%3A58%2C%22args%22%3A%5B%5D%7D
    at new InvariantError2 (http://localhost:3002/node_modules/.vite/deps/@apollo_client.js?v=80e95439:197:26)
chunk-VPTIPXRI.js?v=80e95439:14036 The above error occurred in the <Tracks> component:
    at Tracks (http://localhost:3002/src/pages/tracks.js?t=1733807718822:15:36)

// TracksPage.js

import React from "react";
import { useQuery, gql } from "@apollo/client";
import TrackCard from "../containers/track-card";
import { Layout } from "../components";

// GraphQL Query to fetch tracks
const TRACKS = gql`
  query GetTracks {
    tracksForHome {
      id
      title
      thumbnail
    }
  }
`;
const Tracks = () => {
  const { loading, error, data } = useQuery(TRACKS);
  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error! {error.message}</p>;
  if (!data || !data.tracksForHome) {
    return <p>No tracks found.</p>;
  }
  return (
    <Layout grid>
      <QueryResult error={error} loading={loading} data={data}>
        {data?.tracksForHome?.map((track, index) => (
          <TrackCard key={track.id} track={track} />
        ))}
      </QueryResult>
    </Layout>
  );
};
export default Tracks;

**Reply by testqual** [reply]
The issue is sorted.

**Comment by slayercoder** [comment]
Hey I am following the tutorial but looks like the hook is always in loading state, I checked the graphql client on editor, the query seems to be working correctly there but on react app on localhost:3000 there is a forever loading state, below is the client side code in Tracks component

```
import React from 'react';
import { Layout } from '../components';
import { gql, useQuery } from '@apollo/client';

/** TRACKS query to retrieve all tracks */
const TRACKS = gql`
  query GetTracks {
    tracksForHome {
      id
      title
      author {
        id
        name
        photo
      }
      thumbnail
      length
      modulesCount
    }
  }
`;

/**
 * Tracks Page is the Catstronauts home page.
 * We display a grid of tracks fetched with useQuery with the TRACKS query
 */
const Tracks = () => {
  debugger;
  const { error, loading, data } = useQuery(TRACKS);
  console.log(error, loading, data, 'AAA');
  if (loading) return 'Loading...';

  if (error) return `Error! ${error.message}`;

  return <Layout grid>{JSON.stringify(data)}</Layout>;
};

export default Tracks;

```

**Reply by lizhennessy** [reply]
@slayercoder interesting! Can you see the result come back in the Network tab? If not, curious if any of the queries are working from the frontend.

**Comment by shanefisher1** [comment]
Hello, I get this error for some reason in my useQuery(TRACKS) call, p. My file is identical to the one in the final folder, so not sure what's wrong.

Argument of type 'unknown' is not assignable to parameter of type 'DocumentNode | TypedDocumentNode<any, OperationVariables>'.ts(2345)
const TRACKS: unknown

**Comment by le-el** [comment]
Great!

**Comment by lady-prado** [comment]
I'm getting this error when I try to run the first code challenge:

unknown: Support for the experimental syntax 'jsx' isn't currently enabled (21:10): 19 | if (error) return Error! ${error.message}; 20 |
21 | return <Layout grid>{JSON.stringify(data)}</Layout>; | ^ 22 | };
Add @babel/preset-react (https://github.com/babel/babel/tree/main/packages/babel-preset-react) to the 'presets' section of your Babel config to enable transformation. If you want to leave it as-is, add @babel/plugin-syntax-jsx (https://github.com/babel/babel/tree/main/packages/babel-plugin-syntax-jsx) to the 'plugins' section to enable parsing.

**Reply by mabuyo** [reply]
Hey there @lady-prado! Could you clarify a few things for me:

- Which lesson are you referring to? Can you link to it?
- What was your solution?
- Are you taking the JavaScript or TypeScript version of the course?

This will help us troubleshoot further. Thanks!

**Comment by jleangstartup** [comment]
Your "After all that code, the tracks.js file should look like this: " code is missing this line:

import QueryResult from "../components/query-result";

**Comment by ernestosurribas** [comment]
Great tutorial!

**Comment by jethr0id** [comment]
HI, I've completed everything on this page, but when i run client > npm start, nothing changes on http://localhost:3000/, it is still the same page as before

what else can i provide to get help troubleshooting? thank you!

```
import React from 'react';
import { Layout } from '../components';
import { useQuery, gql } from "@apollo/client";
import TrackCard from "../containers/track-card";
import QueryResult from '../components/query-result';
const TRACKS = gql `
query GetTracks {
  tracksForHome {
    id
    title
    thumbnail
    length
    modulesCount
    author {
      id
      name
      photo
    }
  }
}
`;

/**
 * Tracks Page is the Catstronauts home page.
 * We display a grid of tracks fetched with useQuery with the TRACKS query
 */
const Tracks = () => {
  const { loading, error, data } = useQuery(TRACKS);

  return (
  <Layout grid>
<QueryResult error={error} loading={loading} data={data}>
  {data?.tracksForHome?.map((track) => (
    <TrackCard key={track.id} track={track} />
  ))}
</QueryResult>
</Layout>
);
};

export default Tracks;

```

![2025-06-23_12-57-17](https://github.com/user-attachments/assets/d729dc6d-557f-4475-8f2d-a9446c492404)
<img width="1728" alt="image" src="https://github.com/user-attachments/assets/141bb5cc-120f-4acf-8cd1-696ca5d2400d" />

**Comment by amezir** [comment]
Great tutorial ! I would have liked the videos to have been updated too. But otherwise it's great 🔥

**Comment by valentinowyhnel** [comment]
Great tutorial !
